[LV, VP]VP intrinsics support for the Loop Vectorizer + adding new tail-folding mode using EVL.

alexey-bataev opened the pull request:
This patch introduces generating VP intrinsics in the Loop Vectorizer.

Currently the Loop Vectorizer supports vector predication in a very limited capacity via tail-folding and masked load/store/gather/scatter intrinsics. However, this does not let architectures with active vector length predication support take advantage of their capabilities. Architectures with general masked predication support also can only take advantage of predication on memory operations. By having a way for the Loop Vectorizer to generate Vector Predication intrinsics, which (will) provide a target-independent way to model predicated vector instructions. These architectures can make better use of their predication capabilities.

Our first approach (implemented in this patch) builds on top of the existing tail-folding mechanism in the LV (just adds a new tail-folding mode using EVL), but instead of generating masked intrinsics for memory operations it generates VP intrinsics for loads/stores instructions. The patch adds a new VPlanTransforms to replace the wide header predicate compare with EVL and updates codegen for load/stores to use VP store/load with EVL.

Other important part of this approach is how the Explicit Vector Length is computed. (VP intrinsics define this vector length parameter as Explicit Vector Length (EVL)). We use an experimental intrinsic `get_vector_length`, that can be lowered to architecture specific instruction(s) to compute EVL.

Also, added a new recipe to emit instructions for computing EVL. Using VPlan in this way will eventually help build and compare VPlans corresponding to different strategies and alternatives.

Differential Revision: https://reviews.llvm.org/D99750

alexey-bataev commented:
Ping!

fhahn commented:
Thanks for moving to Github now that Phabricator has been taken down!

 I think @ayalz added some comments shortly before Phabricator was deactivated; unfortunately https://reviews.llvm.org/D99750 isn't accessible at the moment it seems (and it also doesn't seem to be available at http://108.170.204.19/D99750 which is supposed to have a static mirror). I am not sure what's the best way to pick up the recent comments here, perhaps it would be best to share the latest responses here on GH now?

alexey-bataev commented:
I addressed most of the @ayalz  comments in this version

fhahn commented:
> I addressed most of the @ayalz comments in this version

Ok thanks!

 It would be helpful to import the recent conversations here and including what has been addressed how in the current iteration and if anything is still left open. Unfortunately it looks like for some reason D99750  isn't included in the static archive of reviews.llvm.org, I posted https://discourse.llvm.org/t/some-reviews-on-reviews-llvm-org-seem-to-be-missing-from-the-static-archive/76001 to hopefully get back access to the context in Phabricator.

alexey-bataev commented:
Rebase

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -348,6 +348,44 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
+    assert(EVL->getType()->getScalarSizeInBits() <=
+               Phi->getType()->getScalarSizeInBits() &&
+           "EVL type must be smaller than Phi type.");
+    EVL = Builder.CreateIntCast(EVL, Phi->getType(), /*isSigned=*/false);
```
Would it be possible to use the same type for all users without needing to cast here? Without the case, would a simple `Add` VPInstruction suffice (as in a5891fa4d2b76)

alexey-bataev replied:
I tried but it does not work unfortunately. It would be good to have Cast VPRecipe to implement this without adding new Instruction.
The type of the EVL (and many of their users) is i32 (because of https://llvm.org/docs/LangRef.html#llvm-experimental-get-vector-length-intrinsic) and the cast is required

fhahn replied:
Ah yes, I remember now again! There's now a recipe for vector casts, but not yet for scalar casts. Let me check if there are other places that would benefit from such a recipe.

fhahn replied:
Looks link a general recipe for scalar casts would also be helpful in other cases (e.g. truncate of induction steps), shared a patch for discussion: #78113

alexey-bataev replied:
Still does not work:
opt: lib/Transforms/Vectorize/VPlan.cpp:290: llvm::Value *llvm::VPTransformState::get(llvm::VPValue *, unsigned int): Assertion `(isa<VPWidenIntOrFpInductionRecipe>(Def->getDefiningRecipe()) || isa<VPScalarIVStepsRecipe>(Def->getDefiningRecipe()) || isa<VPExpandSCEVRecipe>(Def->getDefiningRecipe())) && "unexpected recipe found to be invariant"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -348,6 +348,44 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
+    return SetVL;
+  }
+  case VPInstruction::ExplicitVectorLengthIVIncrement: {
+    assert(State.UF == 1 && Part == 0 &&
```
asserting `Part = 0` should be sufficient I think (also above)

fhahn commented on llvm/test/Transforms/LoopVectorize/vectorize-vp-intrinsics-interleave.ll:
```diff
@@ -0,0 +1,169 @@
+; RUN: -force-target-supports-scalable-vectors -scalable-vectorization=on -S < %s | FileCheck --check-prefix=NO-VP %s
+
+define void @interleave(ptr noalias %a, ptr noalias %b, ptr noalias %c, i64 %N) {
+; IF-EVL-LABEL: @interleave(
```
Is the goal of the target independent tests to show that no EVL code is generated, as it's not supported? Might be good to add a comment to the tests to clarify.

alexey-bataev replied:
Ok, will add

alexey-bataev commented:
The update regarding AVL/EVL. I missed one point here, when we discussed it before.
1. AVL can be referred as the input parameter for llvm.experimental.get.vector.length intrinsic.
2. EVL is the result, returned by this intrinsic.

So, this 2 subject are separate. For this reason the corresponding parameter in LLVM IR  Reference manual (https://llvm.org/docs/LangRef.html) for VP-based intrinsics is named as `%evl` and we use it here as EVL.

alexey-bataev commented:
Rebase + merged the check lines in the tests

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9414,6 +9475,52 @@ void VPReplicateRecipe::execute(VPTransformState &State) {
+static Instruction *
+lowerStoreUsingVectorIntrinsics(IRBuilderBase &Builder, Value *Addr,
+                                Value *StoredVal, bool IsScatter, Value *Mask,
+                                Value *EVLPart, const Align &Alignment) {
```
dropping `Part` from `EVL` seem still pending (https://reviews.llvm.org/D99750?id=558054#inline-1551817)

fhahn replied:
Still pending?

alexey-bataev replied:
Dropped

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9445,14 +9552,31 @@ void VPWidenMemoryInstructionRecipe::execute(VPTransformState &State) {
       Instruction *NewSI = nullptr;
       Value *StoredVal = State.get(StoredValue, Part);
-      if (CreateGatherScatter) {
+      if (State.EVL) {
```
Still accessing EVL (as per https://reviews.llvm.org/D99750?id=558054#inline-1551722).

To move forward and make progress, I think this would suggest to leave as is now, with a TODO to model this explicitly in the recipe.

I am planning on splitting up the memory recipes soon now that address generation has been moved.

alexey-bataev replied:
Need to check that EVL is accessible, cannot remove it now

ayalz replied:
Above was raised earlier in
https://reviews.llvm.org/D99750#inline-1551721

>> Recipes should strive to have straightforward code-gen as much as possible (contrary to "smart vector instructions/vp intrinsics emission" of the 2nd bullet in the summary's Tentative Development Roadmap).
This is already challenged by the existing (non-EVL) VPWidenMemoryInstructionRecipe::execute.
>>
>> Design dedicated recipe(s) for widening memory instructions under EVL, and introduce them instead of the existing non-EVL recipes, preferably as a VPlan-to-VPlan transformation, rather than try to fit everything here, and potentially elsewhere?
Also discussed below in https://reviews.llvm.org/D99750#inline-967127, and iinm in earlier revisions.

> Why? I t already handles masking, why it should not be extended with the handling of EVL? New recipe will not add anything new here, just will be copy/paste of the existing recipes.

Reason explained above: execute() of recipes should be straightforward.
This is one of the main guidelines outlined in the VPlan roadmap.
This recipe is getting too complicated, should probably separate gather/scatter from wide load/store, and separate the pointer setting (as in [VPlan] Model address separately. #72164), independent of this patch.
Recipe should indicate statically if EVL is used or not, to simplify code-gen and facilitate cost estimation, rather than having to check State.EVL during execute().
If multiple recipes share some common core, it can be shared via a common base class, as in VPHeaderPHIRecipe and VPRecipeWithIRFlags.

alexey-bataev replied:
> Above was raised earlier in https://reviews.llvm.org/D99750#inline-1551721
> 
> > > Recipes should strive to have straightforward code-gen as much as possible (contrary to "smart vector instructions/vp intrinsics emission" of the 2nd bullet in the summary's Tentative Development Roadmap).
> > > This is already challenged by the existing (non-EVL) VPWidenMemoryInstructionRecipe::execute.
> > > Design dedicated recipe(s) for widening memory instructions under EVL, and introduce them instead of the existing non-EVL recipes, preferably as a VPlan-to-VPlan transformation, rather than try to fit everything here, and potentially elsewhere?
> > > Also discussed below in https://reviews.llvm.org/D99750#inline-967127, and iinm in earlier revisions.
> 
> > Why? I t already handles masking, why it should not be extended with the handling of EVL? New recipe will not add anything new here, just will be copy/paste of the existing recipes.
> 
> Reason explained above: execute() of recipes should be straightforward. This is one of the main guidelines outlined in the VPlan roadmap. This recipe is getting too complicated, should probably separate gather/scatter from wide load/store, and separate the pointer setting (as in [VPlan] Model address separately. #72164), independent of this patch. Recipe should indicate statically if EVL is used or not, to simplify code-gen and facilitate cost estimation, rather than having to check State.EVL during execute(). If multiple recipes share some common core, it can be shared via a common base class, as in VPHeaderPHIRecipe and VPRecipeWithIRFlags.

Do you want me to spend another year right now discussing the design of this new class?

fhahn replied:
As mentioned earlier, I think we are in agreement that the recipe should be split up (also independent of EVL support). It's something I am planning to get around to do sometime soon, but there are a number of other changes I would like  to wrap up first.

So I don't think we need to block the current patch on the refactoring, but leave a TODO to split up the recipe in general.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2288,6 +2298,39 @@ class VPActiveLaneMaskPHIRecipe : public VPHeaderPHIRecipe {
+  }
+
+  /// Generate phi for handling IV based on EVL over iterations correctly.
+  void execute(VPTransformState &State) override;
```
> Is a dedicated recipe needed, or would some general header-phi recipe/VPInstruction suffice?
(from https://reviews.llvm.org/D99750?id=558054#inline-1552139)

This would need a scalar phi and could share code with VPCanonicalIVPHIRecipe I think. I think this can be done as cleanup as follow-up, would be good to leave a TODO

alexey-bataev replied:
Added

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -348,6 +348,42 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
+
+      Value *GVL = State.Builder.CreateIntrinsic(
+          State.Builder.getInt32Ty(), Intrinsic::experimental_get_vector_length,
+          {EVL, VFArg, State.Builder.getTrue()});
```
assert that VF is scalable

alexey-bataev replied:
Done!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -348,6 +348,42 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
+    // Compute VTC - IV as the EVL(requested vector length).
+    Value *Index = State.get(getOperand(0), 0);
+    Value *TripCount = State.get(getOperand(1), VPIteration(0, 0));
+    Value *EVL = State.Builder.CreateSub(TripCount, Index);
```
> Suggestion to use`ApplicationVectorLength` (from  https://reviews.llvm.org/D99750?id=558054#inline-1552142)

alexey-bataev replied:
AS I said before, ApplicationVectorLength is just an input argument of the get_vector_length instinsic, while EVL - is the result, returned by this intrinsic. It is used in all LLVM documents with this meaning.

ayalz replied:
So we're in agreement that `TC - Index`, computed here and fed as an input argument to the get_vector_length instinsic, is in fact the Application VL to be named `AVL`, and the result returned by `GetSetVL()` should be named `EVL`, right? Rather than `EVL` and `SetVL`, respectively.

alexey-bataev replied:
Yes

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vplan-vp-intrinsics.ll:
```diff
@@ -0,0 +1,133 @@
+; RUN: -mtriple=riscv64 -mattr=+v -riscv-v-vector-bits-max=128 -disable-output < %s 2>&1 | FileCheck --check-prefixes=NO-VP,CHECK %s
+
+define void @foo(ptr noalias %a, ptr noalias %b, ptr noalias %c, i64 %N) {
+; IF-EVL: VPlan 'Initial VPlan for VF={vscale x 1,vscale x 2,vscale x 4},UF>=1' {
```
> UF>=1 or UF=1? (https://reviews.llvm.org/D99750?id=558054#inline-1551940)

We are quite close to explicit unrolling, so UF should be made explicit soon. Would be good to still indicate the valid range of UFs.

alexey-bataev replied:
Am, not the problem of this patch. This is how the Vplan is printed, separate problem.

fhahn replied:
When introducing the EVL, the transform should do something like `Plan->setUF(1)`

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4690,6 +4712,39 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
   // FIXME: look for a smaller MaxVF that does divide TC rather than masking.
   if (Legal->prepareToFoldTailByMasking()) {
```
> The logic of this method is getting excessively complicated.
> Suggest to turn this into an early-exiting if (!Legal->prepareToFoldTailByMasking()) https://reviews.llvm.org/D99750?id=558054#inline-1551712

alexey-bataev replied:
I simplified it as much as I could. If you have any particular comments, how it can be improved, please be more specific.

fhahn replied:
Early exit on `(!Legal->prepareToFoldTailByMasking())` won't work, because there is other code below.

But code can still be simplified by early exits, something like below should simplify things a bit
```diff
-    if (MaxFactors.ScalableVF.isVector()) {
-      assert(MaxFactors.ScalableVF.isScalable() &&
-             "Expected scalable vector factor.");
-      // FIXME: use actual opcode/data type for analysis here.
-      PreferEVL = getTailFoldingStyle() == TailFoldingStyle::DataWithEVL &&
-                  TTI.hasActiveVectorLength(0, nullptr, Align());
-#if !NDEBUG
-      if (getTailFoldingStyle() == TailFoldingStyle::DataWithEVL) {
-        if (PreferEVL)
-          dbgs() << "LV: Preference for VP intrinsics indicated. Will "
-                    "try to generate VP Intrinsics.\n";
-        else
-          dbgs() << "LV: Preference for VP intrinsics indicated. Will "
-                    "not try to generate VP Intrinsics since the target "
-                    "does not support vector length predication.\n";
-      }
-#endif // !NDEBUG
-
-      // Tail folded loop using VP intrinsics restricts the VF to be scalable
-      // for now.
-      // TODO: extend it for fixed vectors, if required.
-      if (PreferEVL)
-        MaxFactors.FixedVF = ElementCount::getFixed(1);
+    // FIXME: use actual opcode/data type for analysis here.
+    PreferEVL = MaxFactors.ScalableVF.isScalable() &&
+                getTailFoldingStyle() == TailFoldingStyle::DataWithEVL &&
+                TTI.hasActiveVectorLength(0, nullptr, Align());
+    if (!PreferEVL) {
+      LLVM_DEBUG(dbgs() << "LV: Preference for VP intrinsics indicated. Will "
+                           "not try to generate VP Intrinsics since the target "
+                           "does not support vector length predication.\n");
+      return MaxFactors;
     }

+    LLVM_DEBUG(dbgs() << "LV: Preference for VP intrinsics indicated. Will "
+                         "try to generate VP Intrinsics.\n");
+
+    // Tail folded loop using VP intrinsics restricts the VF to be scalable
+    // for now.
+    // TODO: extend it for fixed vectors, if required.
+    assert(MaxFactors.ScalableVF.isScalable() &&
+           "Expected scalable vector factor.");
+    MaxFactors.FixedVF = ElementCount::getFixed(1);
```

fhahn commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.h:
```diff
@@ -75,6 +75,22 @@ class RISCVTTIImpl : public BasicTTIImplBase<RISCVTTIImpl> {
                                       const APInt &Imm, Type *Ty,
                                       TTI::TargetCostKind CostKind);
 
+  /// \name Vector Predication Information
```
There was a long comment re terminology here https://reviews.llvm.org/D99750#inline-1551705, but I cannot really tell how/if this still applies from the context.

alexey-bataev replied:
AS I said before, for 2 different objects we use different names.

fhahn reviewed (commented):
The review on Phabricator is now available on the static archive: https://reviews.llvm.org/D99750

Went through @ayalz 's latest comments and shared the one that still seems pending/open. There was quite a lot to go over, so I might have missed some comments.

Also added some more comments inline. In terms of further refactoring for this patch, would be good to remove the dedicated EVLIncrement opcode now that #78113 landed, if possible.

The other larger pending suggestion is related to EVL handling in the recipes; I suggest to add a TODO and address this as follow-up, unless @ayalz prefers doing the refactoring first. I am planning to split up/refactor memory recipe soon now that address computation is already moved out.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4690,6 +4712,39 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
+             "Expected scalable vector factor.");
+      // FIXME: use actual opcode/data type for analysis here.
+      PreferEVL = getTailFoldingStyle() == TailFoldingStyle::DataWithEVL &&
+                  TTI.hasActiveVectorLength(0, nullptr, Align());
```
The does n to introduce it, it already exists, was introduced some time ago for PowerPC.

fhahn replied:
Oh right, surprised it is already used by PPC. EVL LV won't work on PPC due to them not using scalable vectors? At least I cannot find a test that uses vscale. But would it make sense to support EVL on PPC, if it supports active-vector-length?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -244,6 +244,12 @@ struct VPTransformState {
   unsigned UF;
 
+  /// If EVL is not nullptr, then EVL must be a valid value set during plan
+  /// creation, possibly a default value = whole vector register length. EVL is
```
Yes, just for now, later this should be removed and used as operand explicitly in VPInstructions/recipes.

fhahn replied:
Yep, I meant it looks like the comment needs updating, as it says it is set during plan creation.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -348,6 +348,42 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
+    Value *TripCount = State.get(getOperand(1), VPIteration(0, 0));
+    Value *EVL = State.Builder.CreateSub(TripCount, Index);
+    Value *SetVL = GetSetVL(State, EVL);
+    State.EVL = this;
```
Done.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4690,6 +4712,39 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
+      }
+#endif // !NDEBUG
+
+      // Tail folded loop using VP intrinsics restricts the VF to be scalable.
```
Added TODO to extend for support for fixed vector length.

fhahn reviewed (commented):
> Instruction::Add still does not work, crashes the compiler because this VPInstruction returns that it "does not use only first lane".

Can you push that commit somewhere so I can have a look? Looks like only-first-lane-used analysis might need some additional info.

alexey-bataev commented:
> > Instruction::Add still does not work, crashes the compiler because this VPInstruction returns that it "does not use only first lane".
> 
> Can you push that commit somewhere so I can have a look? Looks like only-first-lane-used analysis might need some additional info.

Just replace VPInstruction::ExplicitVectorLengthIVIncrement with Instruction::Add in lib/Transforms/Vectorize/VPlanTransforms.cpp, line 1273

alexey-bataev commented:
> Oh right, surprised it is already used by PPC. EVL LV won't work on PPC due to them not using scalable vectors? At least I cannot find a test that uses vscale. But would it make sense to support EVL on PPC, if it supports active-vector-length?

Currently it won't work for PPC, since it has some specific checks in TTI. We (or PPC developers) can enable it later.

alexey-bataev commented:
> Yep, I meant it looks like the comment needs updating, as it says it is set during plan creation.

Changed it to "during plan transformation"

alexey-bataev commented:
> When introducing the EVL, the transform should do something like Plan->setUF(1)

It does. It adds UF=1 to Ufs, but VPlan print does not print the list of Ufs, it just checks the size and if UFs is not empty, prints "UF>=1". "UF=1" is printed only if UFs is empty.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1156,6 +1156,45 @@ static VPActiveLaneMaskPHIRecipe *addVPLaneMaskPhiAndUpdateExitBranch(
+    }
+  }
+  if (!WideCanonicalIV->getNumUsers())
+    WideCanonicalIV->eraseFromParent();
```
> > When introducing the EVL, the transform should do something like Plan->setUF(1)
> 
> It does. It adds UF=1 to Ufs, but VPlan print does not print the list of Ufs, it just checks the size and if UFs is not empty, prints "UF>=1". "UF=1" is printed only if UFs is empty.

I might be missing something but the code here doesn't seem to do `setUF(1)`. If `UFs` are set, it should only print those (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/VPlan.cpp#L972)

alexey-bataev replied:
Maybe I read it wrong, will do.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9479,14 +9583,31 @@ void VPWidenMemoryInstructionRecipe::execute(VPTransformState &State) {
       Value *StoredVal = State.get(StoredValue, Part);
-      if (CreateGatherScatter) {
+      if (State.EVL) {
+        Value *EVLPart = State.get(State.EVL, Part);
```
Would be good to assert here that `Part == 0`, same for loads

fhahn replied:
> "Expected only UF==1 for predicated vectorization.");

Predicated vectorization seems ambiguous here, regular masked loads/stores are also predicated. Would probably be better to clearly refer to EVL, something like `Expected only UF == 1 when vectorizing with explicit vector length`.

Drop the part from `EVLPart` here as well? (going back to https://reviews.llvm.org/D99750?id=558054#inline-1551817))

fhahn reviewed (commented):
> Instruction::Add still does not work, crashes the compiler because this VPInstruction returns that it "does not use only first lane".

Ah yes thanks for checking, had a look at this and there's also another issue, that is how some VPInstructions access and store scalar values as `vector` values in `VPTransformState`, which doesn't interact well with other recipes in general. I'll soon share patches to address those 2 shortcomings, but I'd expect that they will take some time to land. Don't think we need to block this patch until that is done, replacing the custom opcode with `::Add` can be done when they land.

The latest version of the patch has diverged quite a bit from the description I think, could you update the title & description to reflect the current scope of the patch?

alexey-bataev commented:
> > Instruction::Add still does not work, crashes the compiler because this VPInstruction returns that it "does not use only first lane".
> 
> Ah yes thanks for checking, had a look at this and there's also another issue, that is how some VPInstructions access and store scalar values as `vector` values in `VPTransformState`, which doesn't interact well with other recipes in general. I'll soon share patches to address those 2 shortcomings, but I'd expect that they will take some time to land. Don't think we need to block this patch until that is done, replacing the custom opcode with `::Add` can be done when they land.
> 
> The latest version of the patch has diverged quite a bit from the description I think, could you update the title & description to reflect the current scope of the patch?

I checked the description, did not find big difference between the description and the code. Could you point explicitly, which part must be updated?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -349,6 +349,42 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
+    return SetVL;
+  }
+  case VPInstruction::ExplicitVectorLengthIVIncrement: {
+    assert(Part == 0 && "Expected unroll factor 1 for VP vectorization.");
```
would be good to leave a TODO here to replace this with a regular Add VPInstruction

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9492,14 +9596,33 @@ void VPWidenMemoryInstructionRecipe::execute(VPTransformState &State) {
     }
   }
 
+  auto MaskValue = [&](unsigned Part) -> Value * {
```
might be simpler to just move `        Value *MaskPart = isMaskRequired ? BlockInMaskParts[Part] : nullptr;` up outside the if?

fhahn reviewed (commented):
>I checked the description, did not find big difference between the description and the code. Could you point explicitly, which part must be updated?

For the title, I'd suggest explicitly stating that this adds a new tail-folding mode using EVL

```
Our first approach (implemented in this patch) builds on top of the existing tail-folding mechanism in the LV, but instead of generating masked intrinsics for memory operations it generates VP intrinsics for loads/stores instructions.
```

Would be good to mention this patch 1) adds a new VPlanTransforms to replace the wide header predicate compare with EVL and 2) updates codegen for load/stores to use VP store/load with EVL.

```
Other important part of this approach is how the Explicit Vector Length is computed. (We use active vector length and explicit vector length interchangeably; VP intrinsics define this vector length parameter as Explicit Vector Length (EVL)). We consider the following three ways to compute the EVL parameter for the VP Intrinsics.

The simplest way is to use the VF as EVL and rely solely on the mask parameter to control predication. The mask parameter is the same as computed for current tail-folding implementation.
The second way is to insert instructions to compute min(VF, trip_count - index) for each vector iteration.
For architectures like RISC-V, which have special instruction to compute/set an explicit vector length, we also introduce an experimental intrinsic get_vector_length, that can be lowered to architecture specific instruction(s) to compute EVL.
```

Better to just state what option is chosen for the implementation (or at least make it explicit that the patch only adds the 3rd one?)

I'd also drop the tentative roadmap and mention the current limitation (UF = 1, no reductions, dep-distance restrictions, no reverse access)

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9529,7 +9647,21 @@ void VPWidenMemoryInstructionRecipe::execute(VPTransformState &State) {
+    if (State.EVL) {
+      assert(State.UF == 1 &&
+             "Expected only UF==1 for predicated vectorization.");
+      Value *EVLPart = State.get(State.EVL, Part);
```
Drop the part from EVLPart here as well? (going back to https://reviews.llvm.org/D99750?id=558054#inline-1551817))

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1186,18 +1225,74 @@ void VPlanTransforms::addActiveLaneMask(
+  EVLPhi->addOperand(NextEVLIV);
+
+  // Replace all uses of VPCanonicalIVPHIRecipe by
+  // VPEVLBasedIVPHIRecipe except for VPInstruction::CanonicalIVIncrement.
```
nit: expect for the canonical IV increment. (VPInstruction::CanonicalIVIncrement has ben replaced by a regular Instruction::Add opcode)

alexey-bataev replied:
Done

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4733,6 +4755,36 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
+                getTailFoldingStyle() == TailFoldingStyle::DataWithEVL &&
+                TTI.hasActiveVectorLength(0, nullptr, Align());
+    if (!PreferEVL) {
+      LLVM_DEBUG(dbgs() << "LV: Preference for VP intrinsics indicated. Will "
```
IIUC if `!PreferEVL` we fall back to `TailFoldingStyle::DataWithoutLaneMask` making the value returned by `getTailFoldingStyle()` incorrect, right? This may be confusing and we also ignore the selected tail folding style due to any restrictions in `useVPIWithVPEVLVectorization`.  

AFAICT this isn't an issue with any of the other tail folding styles, as they can always be used if tail folding is valid. For EVL, I assume that even if the restrictions due to current implementation limitations are dropped, we still may not be able to use EVL, because there may be operations in the loop which don't support active-vector-length?

If that's the case, it would probably be cleaner to store the tail-folding style instead of `PreferEVL` and check the selected style early and adjust it if needed. After this point, it should be enough to check the set tail-folding style.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4733,6 +4755,36 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
+    }
+
+    // FIXME: use actual opcode/data type for analysis here.
+    PreferEVL = MaxFactors.ScalableVF.isScalable() &&
```
Would be good to leave a comment here explaining why this is limited to scalable VFs only for now. After briefly looking at the PPC implementation for `hasActiveVectorLength` we would need to check if all loads & stores are of supported types

Might also be good to have a test for PPC to cover this path

alexey-bataev replied:
1. Added FIXME.
2. Added tests for PPC.

fhahn reviewed (commented):
Thanks for the updates!

A few more comments inline, the only main thing that remains from my side is potentially clarifying the handling of tail-folding styles.

> Currently it won't work for PPC, since it has some specific checks in TTI. We (or PPC developers) can enable it later.

I see, thanks! Left a comment inline.

(If possible it would be good to avoid rebase & force push for updates to the PR as recommended here https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes)

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -349,6 +349,43 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
+      assert(State.VF.isScalable() && "Expected scalable vector factor.");
+      Value *VFArg = State.Builder.getInt32(State.VF.getKnownMinValue());
+
+      Value *GVL = State.Builder.CreateIntrinsic(
```
This intrinsic effectively calls vsetvli, whose first argument should be the Application VL, and whose return value is the Effective VL (or simply VL). Yet the first argument is called `EVL` and the return value `GVL`?

alexey-bataev replied:
Renamed

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:
```diff
@@ -207,14 +207,14 @@ Type *VPTypeAnalysis::inferScalarType(const VPValue *V) {
-              })
+                VPReductionPHIRecipe, VPWidenPointerInductionRecipe,
+                VPEVLBasedIVPHIRecipe>([this](const auto *R) {
+            // Handle header phi recipes, except VPWienIntOrFpInduction
```
```suggestion
            // Handle header phi recipes, except VPWidenIntOrFpInduction
```
nit: worth fixing a typo as we reindent.

ayalz reviewed (commented):
In the process of going over this again, posting a first set of comments to pipeline the work.

The second paragraph of the commit message appears broken at the end.

alexey-bataev commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -190,7 +190,10 @@ enum class TailFoldingStyle {
-  DataAndControlFlowWithoutRuntimeCheck
+  DataAndControlFlowWithoutRuntimeCheck,
+  /// Use predicated EVL instructions for tail-folding.
+  /// Indicates that VP intrinsics should be used if tail-folding is enabled.
```
Removed

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -247,10 +248,12 @@ static cl::opt<TailFoldingStyle> ForceTailFoldingStyle(
+                   "Similar to data-and-control, but remove the runtime check"),
+        clEnumValN(TailFoldingStyle::DataWithEVL, "data-with-evl",
+                   "Use predicated EVL instructions for tail folding if the "
+                   "target supports vector length predication")));
```
Removed

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1633,6 +1634,23 @@ class LoopVectorizationCostModel {
   }
 
+  /// Returns true if VP intrinsics with explicit vector length support should
+  /// be generated in the tail folded loop.
```
Reowked it, now it just checks if getTailFoldingStyle() == TailFoldingStyle::DataWithEVL

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4733,6 +4755,36 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
   // FIXME: look for a smaller MaxVF that does divide TC rather than masking.
   if (Legal->prepareToFoldTailByMasking()) {
     CanFoldTailByMasking = true;
+    if (getTailFoldingStyle() == TailFoldingStyle::None)
```
1. Removed this check, as it is covered later. For now, we can have getTailFoldingStyle() == TailFoldingStyle::None.
2. Done

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4733,6 +4755,36 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
+      return MaxFactors;
+
+    if (UserIC > 1) {
+      LLVM_DEBUG(dbgs() << "LV: Preference for VP intrinsics indicated. Will "
+                           "not generate VP intrinsics since interleave count "
```
Reworked it.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4733,6 +4755,36 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
+
+    // FIXME: use actual opcode/data type for analysis here.
+    PreferEVL = MaxFactors.ScalableVF.isScalable() &&
+                getTailFoldingStyle() == TailFoldingStyle::DataWithEVL &&
```
Reworked this part

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4733,6 +4755,36 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
+                getTailFoldingStyle() == TailFoldingStyle::DataWithEVL &&
+                TTI.hasActiveVectorLength(0, nullptr, Align());
+    if (!PreferEVL) {
+      LLVM_DEBUG(dbgs() << "LV: Preference for VP intrinsics indicated. Will "
```
Not quite. It was checked that TailFoldingStyle::DataWithEVL is preferred, but the legality check did not pass. Reworked.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4733,6 +4755,36 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
+    // TODO: extend it for fixed vectors, if required.
+    assert(MaxFactors.ScalableVF.isScalable() &&
+           "Expected scalable vector factor.");
+    MaxFactors.FixedVF = ElementCount::getFixed(1);
```
Added

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -5342,6 +5394,10 @@ LoopVectorizationCostModel::selectInterleaveCount(ElementCount VF,
   if (!isScalarEpilogueAllowed())
     return 1;
 
+  // Do not interleave if EVL is preferred and no User IC is specified.
```
Added

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8599,6 +8655,8 @@ void LoopVectorizationPlanner::buildVPlansWithVPRecipes(ElementCount MinVF,
             *Plan, CM.getMinimalBitwidths(), PSE.getSE()->getContext());
       VPlanTransforms::optimize(*Plan, *PSE.getSE());
+      if (CM.useVPIWithVPEVLVectorization())
+        VPlanTransforms::addExplicitVectorLength(*Plan);
```
I tried it, explained it here long time ago https://reviews.llvm.org/D99750. It does not work because of some reason (some recipes are removed/transformed too early, so cannot put it later in the pipeline)

fhahn replied:
still open to add TODO?

alexey-bataev committed:
Address comments

alexey-bataev committed:
Adjust comments, lambda name

alexey-bataev committed:
Fix formatting

alexey-bataev commented:
Ping!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4728,6 +4740,47 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
   if (Legal->prepareToFoldTailByMasking()) {
     CanFoldTailByMasking = true;
+    if (getTailFoldingStyle() == TailFoldingStyle::DataWithEVL) {
+      assert(!ChosenTailFoldingStyle &&
```
Better to move this to its own helper (e.g. `selectTailFoldingStyle`), have to called before computeMaxVF and move all logic related to selecting the tail-folding style there, so that `getTailFoldingStyle()` is simplify a return of the selected style? Would be good to do this refactoring in a separate PR

fhahn reviewed (commented):
Thanks for the recent update! Only a further suggestion regarding tail-folding style handling and moving that to a separate NFC PR in preparation for this one

appujee commented:
> Thanks for the recent update! Only a further suggestion regarding tail-folding style handling and moving that to a separate NFC PR in preparation for this one

Please +1 if remaining changes can be done in a separate PR.

@ayalz please update if you have further comments.

fhahn commented:
> > Thanks for the recent update! Only a further suggestion regarding tail-folding style handling and moving that to a separate NFC PR in preparation for this one
> 
> Please +1 if remaining changes can be done in a separate PR.

In case I wasn't clear above, the suggestion is to do the refactoring of tail-folding style selection as separate PR and have this PR depend on the refactoring PR.

alexey-bataev commented:
> > > Thanks for the recent update! Only a further suggestion regarding tail-folding style handling and moving that to a separate NFC PR in preparation for this one
> > 
> > 
> > Please +1 if remaining changes can be done in a separate PR.
> 
> In case I wasn't clear above, the suggestion is to do the refactoring of tail-folding style selection as separate PR and have this PR depend on the refactoring PR.


https://github.com/llvm/llvm-project/pull/81885

alexey-bataev committed:
Rebase

alexey-bataev committed:
[LV, VP]VP intrinsics support for the Loop Vectorizer + adding new
tail-folding mode using EVL.

This patch introduces generating VP intrinsics in the Loop Vectorizer.

Currently the Loop Vectorizer supports vector predication in a very limited capacity via tail-folding and masked load/store/gather/scatter intrinsics. However, this does not let architectures with active vector length predication support take advantage of their capabilities. Architectures with general masked predication support also can only take advantage of predication on memory operations. By having a way for the Loop Vectorizer to generate Vector Predication intrinsics, which (will) provide a target-independent way to model predicated vector instructions, These architectures can make better use of their predication capabilities.

Our first approach (implemented in this patch) builds on top of the existing tail-folding mechanism in the LV (just adds a new tail-folding mode using EVL), but instead of generating masked intrinsics for memory operations it generates VP intrinsics for loads/stores instructions. The patch adds a new VPlanTransforms to replace the wide header predicate compare with EVL and updates codegen for load/stores to use VP store/load with EVL.

Other important part of this approach is how the Explicit Vector Length is computed. (VP intrinsics define this vector length parameter as Explicit Vector Length (EVL)). We use an experimental intrinsic `get_vector_length`, that can be lowered to architecture specific instruction(s) to compute EVL.

Also, added a new recipe to emit instructions for computing EVL. Using VPlan in this way will eventually help build and compare VPlans corresponding to different strategies and alternatives.

Differential Revision: https://reviews.llvm.org/D99750

alexey-bataev committed:
Rebase

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1209,20 +1248,75 @@ void VPlanTransforms::addActiveLaneMask(
+    OpVPEVL->insertBefore(CanonicalIVIncrement);
   }
+  auto *NextEVLIV = new VPInstruction(
+      VPInstruction::ExplicitVectorLengthIVIncrement, {OpVPEVL, EVLPhi},
```
I *think* all changes to allow using `Instruction::Add` here should have landed now.

alexey-bataev committed:
Removed ExplicitVectorLengthIVIncrement, replaced by Instruction::Add

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -5256,6 +5309,13 @@ LoopVectorizationCostModel::selectInterleaveCount(ElementCount VF,
     return 1;
 
+  // Do not interleave if EVL is preferred and no User IC is specified.
+  if (useVPIWithVPEVLVectorization()) {
```
nit: more explicit to check tail-folding style directly here or rename helper to something like `foldTailWithEVL`?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -92,7 +92,47 @@ static bool verifyVPBasicBlock(const VPBasicBlock *VPBB,
+  // recipes that must appear at most once in Entry or Exiting blocks.
+  DenseSet<unsigned> EVLFound;
+  const VPlan *Plan = VPBB->getPlan();
+  bool IsHeader = Plan && Plan->getEntry()->getNumSuccessors() == 1 &&
```
Plan must always be set here I think? Why do we need to check the number of successors here?

alexey-bataev replied:
I answered this question before, some of the unit tests crash without this check, they have recipes/vp-blocks without plan

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -92,7 +92,47 @@ static bool verifyVPBasicBlock(const VPBasicBlock *VPBB,
+    switch (RInst->getOpcode()) {
+    case VPInstruction::ExplicitVectorLength:
+      if (!IsHeader) {
+        errs() << "EVL instruction not in entry block!\n";
```
nit: not in header block

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -92,7 +92,47 @@ static bool verifyVPBasicBlock(const VPBasicBlock *VPBB,
+    auto *RInst = dyn_cast<VPInstruction>(R);
+    if (!RInst)
+      return true;
+    switch (RInst->getOpcode()) {
```
simpler to use an if for a single case?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -92,7 +92,47 @@ static bool verifyVPBasicBlock(const VPBasicBlock *VPBB,
     RecipeNumbering[&R] = Cnt++;
 
+  // Set of recipe types along with VPInstruction Opcodes of all EVL-related
+  // recipes that must appear at most once in Entry or Exiting blocks.
```
Entry -> header of the vector loop region?

Drop reference to Exiting.

I think it would also be good to check if all recipes that implicitly use EVL are dominated by the EVL recipe.

fhahn reviewed (commented):
Thanks for the latest updates! Left a few additional comments mostly for the verifier. 

I'd suggest to give @ayalz  a bit more time (perhaps till end of next week or so) in case he has additional thoughts.

alexey-bataev committed:
Address comments

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9430,7 +9552,21 @@ void VPWidenMemoryInstructionRecipe::execute(VPTransformState &State) {
+    if (State.EVL) {
+      assert(State.UF == 1 && "Expected only UF == 1 when vectorizing with "
+                              "explicit vector length.");
+      Value *EVL = State.get(State.EVL, VPIteration(0, 0));
```
Would be good to assert that State.EVL is an EVL VPInstruction here and above, unless that's done elsewhere already

alexey-bataev replied:
Already done.

alexey-bataev committed:
Address comments, rebase

alexey-bataev commented:
Ping!

alexey-bataev commented:
Ping!

appujee commented:
@alexey-bataev @fhahn 

As Ayal has not responded to the patch in weeks (https://github.com/llvm/llvm-project/pull/76172#issuecomment-1942735278) even after pinging, should we follow up in a post commit review if and when Ayal has feedback?

ayalz reviewed (commented):
Thanks for addressing previous comments, adding mostly minor nits.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -242,6 +242,12 @@ struct VPTransformState {
+  /// transformation, possibly a default value = whole vector register length.
+  /// EVL is created only if TTI prefers predicated vectorization, thus if EVL
+  /// is not nullptr it also implies preference for predicated vectorization.
+  VPValue *EVL = nullptr;
```
The canonical IV is implicitly associated with its loop region, by its guaranteed placement as first recipe in the header (dominates all recipes in the region).  What makes this tricker for EVL is that the EVL VPInstruction is placed after the phis, so does not necessarily dominate all recipes.

Once the memory recipes have been split up, this likely can be removed, so might also be good to leave a corresponding TODO here in the meantime?

alexey-bataev replied:
Added TODO to remove this once the recipes properly handle EVL.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1201,20 +1240,75 @@ void VPlanTransforms::addActiveLaneMask(
+
+  // Walk users of WideCanonicalIV and replace all compares of the form
+  // (ICMP_ULE, WideCanonicalIV, backedge-taken-count) with an
+  // all-true-mask.
```
> Does the EVLPhi->OpVPEVL->Next cycle have any users but itself? Such cycles may be subject to DCE'd.

Ah, this is fine, missed that "Replace all uses of VPCanonicalIVPHIRecipe by VPEVLBasedIVPHIRecipe except for the canonical IV increment" part.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1201,20 +1240,75 @@ void VPlanTransforms::addActiveLaneMask(
+  EVLPhi->addOperand(NextEVLIV);
+
+  // Replace all uses of VPCanonicalIVPHIRecipe by
+  // VPEVLBasedIVPHIRecipe except for the canonical IV increment..
```
```suggestion
  // VPEVLBasedIVPHIRecipe except for the canonical IV increment.
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1201,20 +1240,75 @@ void VPlanTransforms::addActiveLaneMask(
+  // all-true-mask.
+  Value *TrueMask =
+      ConstantInt::getTrue(CanonicalIVPHI->getScalarType()->getContext());
+  VPValue *VPTrueMask = Plan.getVPValueOrAddLiveIn(TrueMask);
```
The mask still should be emitted, so we need to store the mask within the recipe. Since it is supposed to split VPWidenMemoryRecipe into multiple classes in future, I suppose it should be fine. For EVL-predicate recipe we still need the mask, even if it is all-true.

topperc replied:
Can you use the recently added `splat` syntax?

From Langref

> When creating a vector whose elements have the same constant value, the preferred syntax is splat (<Ty> Val). For example: “splat (i32 11)”. These vector constants must have [vector type](https://llvm.org/docs/LangRef.html#t-vector) with an element type that matches the splat operand.

alexey-bataev replied:
We're using scalar `true` here. Looks like it requires changes in IR builder (??), which is not part of this patch.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1201,20 +1240,75 @@ void VPlanTransforms::addActiveLaneMask(
+                       /*NumBits=*/32)) {
+    OpVPEVL = new VPScalarCastRecipe(Instruction::ZExt, OpVPEVL,
+                                     CanonicalIVPHI->getScalarType());
+    OpVPEVL->insertBefore(CanonicalIVIncrement);
```
> Why depend on CanonicalIVIncrement, can simply place the cast and the bump after VPEVL in the header?

The increment is placed after CanonicalIVIncrement, so just want to keep ext close to next evl.

> 
> Canonical IV must have a type no smaller than i32?
> 
> Simply ask for the size of Canonical IV's scalar type, than compare if equal to i32?

Done.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1201,20 +1240,75 @@ void VPlanTransforms::addActiveLaneMask(
   }
+  auto *NextEVLIV =
+      new VPInstruction(Instruction::Add, {OpVPEVL, EVLPhi},
+                        {CanonicalIVIncrement->hasNoUnsignedWrap(),
+                         CanonicalIVIncrement->hasNoSignedWrap()},
```
EVL is just a counter, so the flags are the same as for the regular IV, no difference. Or do you have something else in mind?

alexey-bataev committed:
Rebase, address comments

alexey-bataev commented:
Ping!

alexey-bataev commented:
Ping!

alexey-bataev commented:
Ping!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -5256,6 +5309,13 @@ LoopVectorizationCostModel::selectInterleaveCount(ElementCount VF,
 
+  // Do not interleave if EVL is preferred and no User IC is specified.
+  if (foldTailWithEVL()) {
+    LLVM_DEBUG(dbgs() << "LV: Preference for VP intrinsics indicated. "
```
Still open?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2445,6 +2452,44 @@ class VPActiveLaneMaskPHIRecipe : public VPHeaderPHIRecipe {
+/// incremented by EVL in each iteration of the vector loop.
+class VPEVLBasedIVPHIRecipe : public VPHeaderPHIRecipe {
+public:
+  VPEVLBasedIVPHIRecipe(VPValue *StartMask, DebugLoc DL)
```
still open?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2445,6 +2452,44 @@ class VPActiveLaneMaskPHIRecipe : public VPHeaderPHIRecipe {
 };
 
+/// A recipe for generating the phi node for the current index of elements,
+/// adjusted in accordance with EVL value. It starts at StartIV value and gets
```
Not sure how doxygen treats `\p` at class definition. Might be better to replace `StartIV` with `the start value of the canonical induction`?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2445,6 +2452,44 @@ class VPActiveLaneMaskPHIRecipe : public VPHeaderPHIRecipe {
+  ~VPEVLBasedIVPHIRecipe() override = default;
+
+  VPEVLBasedIVPHIRecipe *clone() override {
+    return new VPEVLBasedIVPHIRecipe(getOperand(0), getDebugLoc());
```
still pending

alexey-bataev replied:
Done

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -350,6 +350,33 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
+    // TODO: Restructure this code with an explicit remainder loop, vsetvli can
+    // be outside of the main loop.
+    assert(Part == 0 && "No unrolling expected for predicated vectorization.");
+    // Compute VTC - IV as the EVL(requested vector length).
```
still pending?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1172,6 +1172,45 @@ static VPActiveLaneMaskPHIRecipe *addVPLaneMaskPhiAndUpdateExitBranch(
 
+/// Replaces (ICMP_ULE, WideCanonicalIV, backedge-taken-count) pattern using
+/// the given idiom \p Idiom.
+static void replaceHeaderPredicateWithIdiom(
```
still pending?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1201,20 +1240,75 @@ void VPlanTransforms::addActiveLaneMask(
+/// vector.body:
+/// ...
+/// %P = EXPLICIT-VECTOR-LENGTH-BASED-IV-PHI [ %StartV, %vector.ph ], [
+/// %NextEVL, %vector.body ] %EVL = EXPLICIT-VECTOR-LENGTH %P, original TC
```
At least partly still pending?

fhahn reviewed (commented):
Looks like there still are several pending comments, tried to bump them

alexey-bataev committed:
Address pending comments

alexey-bataev commented:
Ping!

alexey-bataev commented:
Ping!

appujee commented:
reviewers@ do you have pending feedback that still needs to be addressed?

appujee commented:
If there aren't pending feedback i'd like to see this merged as we have a dependency on this work.

ayalz reviewed (commented):
I'm OK, pending Florian's approval.
It unfortunately leaves several clean-ups to be addressed as TODO follow-ups.
Added several minor suggestions for improvement.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1505,20 +1508,60 @@ class LoopVectorizationCostModel {
   TailFoldingStyle getTailFoldingStyle(bool IVUpdateMayOverflow = true) const {
-    return IVUpdateMayOverflow ? ChosenTailFoldingStyle.first
-                               : ChosenTailFoldingStyle.second;
+    if (!ChosenTailFoldingStyle.first) {
+      assert(!ChosenTailFoldingStyle.second &&
+             "Chosen tail folding style must not be set.");
+      return TailFoldingStyle::None;
+    }
```
Will try

alexey-bataev replied:
Ubnfortunately, cannot do it directly in the patch, requires too many changes in the code and will definitely cause extra time for the review. Reverting it back.

alexey-bataev commented:
> There are two potential cases where the canonical IV increment may wrap: if the original TC=BTC+1 wraps, and if rounding-up TC to a multiple of VF*UF as part of tail folding wraps. When using EVL the computation BTC+1 must not wrap because EVL uses TC explicitly, and EVL avoids any rounding-up of TC. So does OpVPEVL+EVLPhi not wrap, regardless of CanonicalIVIncrement?
In any case, emitIterationCountCheck() seems to currently ensure that neither the canonical IV increment nor the EVL increment may wrap, when tail folding with EVL.

Why depend on CanonicalIVIncrement, can simply place the cast and the bump after VPEVL in the header?

The increment is placed after CanonicalIVIncrement, so just want to keep ext close to next evl.

It would be good for EVL to be independent of CanonicalIVIncrement, e.g., allowing the latter to be introduced late in the VPlan-to-VPlan pipeline (see https://github.com/llvm/llvm-project/pull/82270).

Currently it it maybe considered independent, but there are some corner cases in future, where we need to copy it from IV. Kolya already highlighted the issue with the later materialization of loop IV.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9387,7 +9507,25 @@ void VPWidenMemoryInstructionRecipe::execute(VPTransformState &State) {
+        // is not nullptr it also implies preference for predicated
+        // vectorization.
+        // FIXME: Support reverse store after vp_reverse is added.
+        Value *MaskPart = isMaskRequired ? BlockInMaskParts[Part] : nullptr;
```
For consecutive non-masked loads it returns nullptr, VectorBuilder then handles it and generates all-true mask.

alexey-bataev commented:
> Thanks @fhahn for answering the "Is there a plan to remove it later?" question, constructively! The excepts from D99750 (w/o response there, sigh) were provided as context.

I answered all questions in D99750, unfortunately some of my latest answers are lost, the backup does not have the latest version.
Yes, we plan to remove it and make it part of recipes, we answered this since the very beginning.

alexey-bataev committed:
Rebase, address comments.

alexey-bataev committed:
Fix formatting

alexey-bataev commented:
Ping!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -242,6 +242,14 @@ struct VPTransformState {
   ElementCount VF;
   unsigned UF;
 
+  /// If EVL is not nullptr, then EVL must be a valid value set during plan
```
nit: would probably be good to mention here what EVL stands for.

fhahn reviewed (commented):
Looks like 2 comments still pending, but that should be it AFAICT

alexey-bataev committed:
Address comments

fhahn commented on llvm/test/Transforms/LoopVectorize/vectorize-vp-intrinsics-interleave.ll:
```diff
@@ -0,0 +1,168 @@
+
+for.body:
+  %iv = phi i64 [ 0, %entry ], [ %iv.next, %for.body ]
+  %arrayidx = getelementptr inbounds i32, ptr %b, i64 %iv
```
Don't think an interleave group will be formed here, as all indices are for different base pointers;

I think it would need something like

```
+  %arrayidx = getelementptr inbounds {i32, i32}, ptr %b, i64 %iv, i32 0
+  %arrayidx2 = getelementptr inbounds {i32, i32}, ptr %b, i64 %iv, i32 1
```

fhahn commented on llvm/test/Transforms/LoopVectorize/vectorize-vp-intrinsics.ll:
```diff
@@ -0,0 +1,98 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```
should this pass `-force-vector-width`? at the moment we vectorize with VF = vscale x 1 and I think this is causing a UBSan failure the running the following tests

```
 LLVM :: Transforms/LoopVectorize/vectorize-vp-intrinsics-iv32.ll
  LLVM :: Transforms/LoopVectorize/vectorize-vp-intrinsics.ll
  LLVM :: Transforms/LoopVectorize/vplan-vp-intrinsics.ll
```

fhahn replied:
nit: better rename to vectorize-force-tail-with-evl.ll as no vp-intrinsics are generated ore referenced in the test

fhahn commented on llvm/test/Transforms/LoopVectorize/vectorize-vp-intrinsics-iv32.ll:
```diff
@@ -0,0 +1,85 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```
nit: better rename to vectorize-force-tail-with-evl.ll as no vp-intrinsics are generated ore referenced in the test

fhahn commented on llvm/test/Transforms/LoopVectorize/vectorize-vp-intrinsics-reverse-load-store.ll:
```diff
@@ -0,0 +1,49 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```
nit: better rename to vectorize-force-tail-with-evl.ll as no vp-intrinsics are generated ore referenced in the test

fhahn commented on llvm/test/Transforms/LoopVectorize/vectorize-vp-intrinsics-no-masking.ll:
```diff
@@ -0,0 +1,36 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```
nit: better rename to vectorize-force-tail-with-evl.ll as no vp-intrinsics are generated ore referenced in the test

fhahn commented on llvm/test/Transforms/LoopVectorize/vplan-vp-intrinsics.ll:
```diff
@@ -0,0 +1,37 @@
+; REQUIRES: asserts
```
nit: better rename to vectorize-force-tail-with-evl.ll as no vp-intrinsics are generated ore referenced in the test

fhahn commented on llvm/test/Transforms/LoopVectorize/vectorize-vp-intrinsics-masked-loadstore.ll:
```diff
@@ -0,0 +1,60 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```
nit: better rename to vectorize-force-tail-with-evl.ll as no vp-intrinsics are generated ore referenced in the test

fhahn reviewed (commented):
It looks like there's a UBSan failure when running some of the tests (comment inline, looks unrelated to the patch, but the tests need to force a vector width > 1), would be good to check with an UBSan build.

Also, the target-independent tests don't exercise the EVL code path due to missing support for VP intrinsics at the moment AFACIT; I think we need to have them with `RISCV` targets as well, as otherwise some code paths aren't exercised in the EVL code path (e.g. interleave groups, gather/scatter, reverse memory ops, masked ops)

fhahn commented:
As for the target-independent tests, would it be sufficient to have a single one to check that EVL doesn't kick in for the simplest case?

alexey-bataev commented:
> As for the target-independent tests, would it be sufficient to have a single one to check that EVL doesn't kick in for the simplest case?

Some time ago (phab) you asked to prepare a test for each particular case. I prepared single file test. You asked to split in multiple files. Did it. Should I again move all these into a single file?

alexey-bataev commented:
> It looks like there's a UBSan failure when running some of the tests (comment inline, looks unrelated to the patch, but the tests need to force a vector width > 1), would be good to check with an UBSan build.

Possible, will double check.

> Also, the target-independent tests don't exercise the EVL code path due to missing support for VP intrinsics at the moment AFACIT; I think we need to have them with `RISCV` targets as well, as otherwise some code paths aren't exercised in the EVL code path (e.g. interleave groups, gather/scatter, reverse memory ops, masked ops)

I had all these tests under RISCV. You asked to make them target independent. Should I again move them to RISCV?

fhahn commented:
> > It looks like there's a UBSan failure when running some of the tests (comment inline, looks unrelated to the patch, but the tests need to force a vector width > 1), would be good to check with an UBSan build.
> 
> Possible, will double check.
> 
> > Also, the target-independent tests don't exercise the EVL code path due to missing support for VP intrinsics at the moment AFACIT; I think we need to have them with `RISCV` targets as well, as otherwise some code paths aren't exercised in the EVL code path (e.g. interleave groups, gather/scatter, reverse memory ops, masked ops)
> 
> I had all these tests under RISCV. You asked to make them target independent. Should I again move them to RISCV?

I'd just move the individual files to RISCV. Earlier comment was if this would retain test coverage (i.e. by adding a something similar like `-force-target-supports-scalable-vectors` for vector-predication), sorry that wasn't clearer from the previous comment

alexey-bataev commented:
> > > It looks like there's a UBSan failure when running some of the tests (comment inline, looks unrelated to the patch, but the tests need to force a vector width > 1), would be good to check with an UBSan build.
> > 
> > 
> > Possible, will double check.
> > > Also, the target-independent tests don't exercise the EVL code path due to missing support for VP intrinsics at the moment AFACIT; I think we need to have them with `RISCV` targets as well, as otherwise some code paths aren't exercised in the EVL code path (e.g. interleave groups, gather/scatter, reverse memory ops, masked ops)
> > 
> > 
> > I had all these tests under RISCV. You asked to make them target independent. Should I again move them to RISCV?
> 
> I'd just move the individual files to RISCV. Earlier comment was if this would retain test coverage (i.e. by adding a something similar like `-force-target-supports-scalable-vectors` for vector-predication), sorry that wasn't clearer from the previous comment

And I answered that all these tests are target-specific, but you still insisted on making them independent. What's changed?

alexey-bataev committed:
Address comments.
1. Fixed UB in areRuntimeChecksProfitable.
2. Moved most of the test to RISCV directory.
3. Fixed a check for supported EVL mode in foldTailWithEVL. Have to
   check later, since the decisions are detected later than
   setTailFoldingStyles().
4. Fixed a crash with VPDerivedIVRecipe.
5. Rebased.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2518,6 +2567,10 @@ class VPDerivedIVRecipe : public VPSingleDefRecipe {
   /// for floating point inductions.
   const FPMathOperator *FPBinOp;
 
+  VPCanonicalIVPHIRecipe *getCanonicalIV() const {
```
After this patch, it/s not guaranteed that the second operand is the canonical IV. Better to completely remove the helper. Also, the documentation for the recipe needs updating above.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1688,8 +1740,8 @@ class LoopVectorizationCostModel {
   /// the IV update may overflow, the second element - if it does not.
-  std::pair<TailFoldingStyle, TailFoldingStyle> ChosenTailFoldingStyle =
-      std::make_pair(TailFoldingStyle::None, TailFoldingStyle::None);
+  std::pair<std::optional<TailFoldingStyle>, std::optional<TailFoldingStyle>>
```
simpler to shift optional outside, if possible, i.e. both are either set or not set?

fhahn reviewed (commented):
Thanks for the update!

> And I answered that all these tests are target-specific, but you still insisted on making them independent. What's changed?


I took another look at the tests after seeing the UBSan failures, which lead to the comments. 
From the previous comments my understanding was that we either have both target-dependent and target-independent versions of the tests or have a flag to force EVL support available (like it exists for forcing scalable vectors available) to have target-independent tests for most checks and target-dependent tests mostly for checking TTI hooks + cost-modeling. 

I don’t remember insisting, just asking if possible with suggestion to add a force flag.

https://reviews.llvm.org/D99750#inline-1551797
> Can this test be target independent? does it need to check the no VP case?
>> 1. No
>> 2. Yes, need to check that the option works correctly
>>> Added also target independent version


https://reviews.llvm.org/D99750#inline-1551796
> This test file is getting quite big with 3 different run lines. I think it would be good to try to split this up a bit, to make it easier to see what's going on.
>
> I'd recommend having the test cases for various legality issues as target-independent tests with force flags (force EVL support, VF and IC). And keep cost-model specific tests target specific.
>> Done

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9584,7 +9743,7 @@ static bool areRuntimeChecksProfitable(GeneratedRTChecks &Checks,
 
   // The scalar cost should only be 0 when vectorizing with a user specified VF/IC. In those cases, runtime checks should always be generated.
-  double ScalarC = *VF.ScalarCost.getValue();
+  uint64_t ScalarC = *VF.ScalarCost.getValue();
```
I assume it should work even with vscale x 1. Plus, I think we'd better to use ints instead of double in the compiler. I prepare a separate patch with this fix.

alexey-bataev committed:
Address comments

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2511,8 +2560,7 @@ class VPWidenCanonicalIVRecipe : public VPSingleDefRecipe {
 /// A recipe for converting the canonical IV value to the corresponding value of
-/// an IV with different start and step values, using Start + CanonicalIV *
-/// Step.
+/// an IV with different start and step values, using Start + IV * Step.
```
The line above also needs updating. Technically the operands now don't have to be IVs (which typically should have a constant step), it just adjust an input now.

fhahn reviewed (approved):
LGTM, thanks!  Would be good to wait another day or so with committing, in case there are any other final comments.

One remaining comment inline for a comment that still needs updating.

I also had time to put up a PR to start with refactoring VPWidenMemoryInstructionRecipe as discussed earlier:  https://github.com/llvm/llvm-project/pull/87411

alexey-bataev committed:
Address comments

vntkmr commented:
@alexey-bataev I just noticed that this patch is merged, good work!
As you might remember, I was the original author of this patch [back in 2021 while at BSC](https://lists.llvm.org/pipermail/llvm-dev/2021-April/149580.html), before you took over. 

I noticed that because of the move from phabricator to github, some of the history of this patch is now lost - I can't find a way to access the initial commits and discussions on the patch from 2021. 
I will appreciate if you can add an acknowledgement for the initial work on this patch.

CC: @rofirrim @simoll

alexey-bataev commented:
> @alexey-bataev I just noticed that this patch is merged, good work! As you might remember, I was the original author of this patch [back in 2021 while at BSC](https://lists.llvm.org/pipermail/llvm-dev/2021-April/149580.html), before you took over.
> 
> I noticed that because of the move from phabricator to github, some of the history of this patch is now lost - I can't find a way to access the initial commits and discussions on the patch from 2021. I will appreciate if you can add an acknowledgement for the initial work on this patch.
> 
> CC: @rofirrim @simoll
> 
> Vineet Kumar

Hi, sure! Sorry, forgot about that :(

alexey-bataev commented:
> @alexey-bataev I just noticed that this patch is merged, good work! As you might remember, I was the original author of this patch [back in 2021 while at BSC](https://lists.llvm.org/pipermail/llvm-dev/2021-April/149580.html), before you took over.
> 
> I noticed that because of the move from phabricator to github, some of the history of this patch is now lost - I can't find a way to access the initial commits and discussions on the patch from 2021. I will appreciate if you can add an acknowledgement for the initial work on this patch.
> 
> CC: @rofirrim @simoll
> 
> Vineet Kumar

Added co-authors to the commit

vntkmr commented:
> > @alexey-bataev I just noticed that this patch is merged, good work! As you might remember, I was the original author of this patch [back in 2021 while at BSC](https://lists.llvm.org/pipermail/llvm-dev/2021-April/149580.html), before you took over.
> > I noticed that because of the move from phabricator to github, some of the history of this patch is now lost - I can't find a way to access the initial commits and discussions on the patch from 2021. I will appreciate if you can add an acknowledgement for the initial work on this patch.
> > CC: @rofirrim @simoll
> > Vineet Kumar
> 
> Hi, sure! Sorry, forgot about that :(

Thank you! 

For those interested in the initial version and older discussion on this patch, it is recorded on the internet archive at https://web.archive.org/web/20230128111909/https://reviews.llvm.org/D99750.
"Show Older Changes" link on the archived phabricator page does not work anymore, unfortunately.

appujee commented:
> > > @alexey-bataev I just noticed that this patch is merged, good work! As you might remember, I was the original author of this patch [back in 2021 while at BSC](https://lists.llvm.org/pipermail/llvm-dev/2021-April/149580.html), before you took over.
> > > I noticed that because of the move from phabricator to github, some of the history of this patch is now lost - I can't find a way to access the initial commits and discussions on the patch from 2021. I will appreciate if you can add an acknowledgement for the initial work on this patch.
> > > CC: @rofirrim @simoll
> > > Vineet Kumar
> > 
> > 
> > Hi, sure! Sorry, forgot about that :(
> 
> Thank you!
> 

Hi Vineet,
Thanks for your work!


> For those interested in the initial version and older discussion on this patch, it is recorded on the internet archive at https://web.archive.org/web/20230128111909/https://reviews.llvm.org/D99750. "Show Older Changes" link on the archived phabricator page does not work anymore, unfortunately.

What you can do is create a PR(on top of the correct baseline) with this change and close the PR. That way it can be there on github as well. It might take a bit but i feel this would help others keep historical context.

appujee commented:
Posted on https://lists.riscv.org/g/sig-toolchains/message/678 notifying interested parties..

fhahn commented:
Are there any plans on adding upstream runtime testing for EVL vectorization to guard against regressions? 

We really should have upstream end-to-end testing that enables the EVL vectorization path and does a stage2 build + llvm-test-suite to catch regressions (similar to how SVE enabled bots were added when scalable vector support was added IIRC)

cc'ing some additional people who might also be able to help @appujee  @Mel-Chen @nikolaypanchenko @arcbbb @preames

appujee commented:
> Are there any plans on adding upstream runtime testing for EVL vectorization to guard against regressions?
> 
> We really should have upstream end-to-end testing that enables the EVL vectorization path and does a stage2 build + llvm-test-suite to catch regressions (similar to how SVE enabled bots were added when scalable vector support was added IIRC)
> 
> cc'ing some additional people who might also be able to help @appujee @Mel-Chen @nikolaypanchenko @arcbbb @preames

+1. Let me know if i can help with anything here.

asb commented:
> Are there any plans on adding upstream runtime testing for EVL vectorization to guard against regressions?
> 
> We really should have upstream end-to-end testing that enables the EVL vectorization path and does a stage2 build + llvm-test-suite to catch regressions (similar to how SVE enabled bots were added when scalable vector support was added IIRC)

I'm currently working through a project to spin up a range of RISC-V builders. Part of that will involve deciding what configurations to test given the resources we have. It sounds like this could be an interesting config to add to the list. To clarify, is the suggestion basically a build with `-mllvm -force-tail-folding-style=data-with-evl`?

alexey-bataev commented:
Hi Alex, yes, at least one builder should enable this option. I think we need to test both configs for now, with and without this option.

topperc commented:
> data-with-evl

Do we also need `-prefer-predicate-over-epilogue=predicate-dont-vectorize`?
